server: rolling restart planning endpoint

MattWhelan opened the pull request:
Iterates over groups of nodes that can be drained and restarted in parallel without creating unavailability.

Epic: none
Part of: https://cockroachlabs.atlassian.net/browse/TREQ-929

Release node (ops change):
The drain planning endpoint aids in efficient rolling restarts by examining replica placement and range health, and identifying non-overlapping nodes that can be drained and restarted without creating unavailability. It also respects cluster capacity constraints.

Commit:
Rolling restart planning endpoint

Iterates over groups of nodes that can be drained and restarted in
parallel without creating unavailability.

Epic: none
Part of: https://cockroachlabs.atlassian.net/browse/TREQ-929

Release node (ops change):
The drain planning endpoint aids in efficient rolling restarts by
examining replica placement and range health, and identifying
non-overlapping nodes that can be drained and restarted without creating
unavailability. It also respects cluster capacity constraints.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/143246)
<!-- Reviewable:end -->

Commit:
Rename DTO

Commit:
Address FP rounding error in batch sizes

Commit:
Require targetCapacity

Commit:
Rate limiting

Commit:
Graph-based approach to rolling restart planning

After realizing that the node to range map was going to be very large,
it occurred to me to approach this as a graph problem. Cluster nodes are
nodes, and they're connected to other nodes when they share at least one
range. Nodes can be restarted in parallel when they don't share an edge.
So we just have to remove each candidate and all its neighbors from the
candidate set.

This commit adds a way to get the termination safety of a node, along
with its neighbors. Neighbor-finding is easy locally, and the size of
the node connection graph is relatively small, compared to the count of
ranges.

MattWhelan commented:
Closing in favor of #150673
